experimental/tests/test_seq_kv_cache.py

```diff
@@ -0,0 +1,313 @@
+
+# Actual
+{micros_per_mb=:.1f}µs per microbatch
+{micros_per_seq=:.1f}µs per seq
+
+{micros_per_seq/expected_micros_per_seq:.1f}x the expected HBM-bandwidth bound time
+"""
+    )
+
+
+def test_garbage_pad_active_queries_throughput(n_active_ctx_per_seq=5):
+    n_seqs = 100
+    d_model_per_gpu = 12 * 1024 // 8
+    active_queries = RaggedActivations.from_list(
+        [
+            torch.ones(n_active_ctx_per_seq, d_model_per_gpu, **bf16_cuda()) * 2
+            for _ in range(n_seqs)
+        ]
+    )
+
+    bytes_in_queries_per_seq = n_active_ctx_per_seq * d_model_per_gpu * 2  # 2 from bf16
+    bytes_in_queries_total = bytes_in_queries_per_seq * n_seqs
+    hbm_bw_bytes_per_gpu = 1555e9  # 1.5TB/s
+
+    # If we just read the bytes directly from memory
+    theor_load_micros_per_seq = bytes_in_queries_per_seq / hbm_bw_bytes_per_gpu * 1e6
+
+    # Doing our operation should be slower than the theoretical minimum because we
+    # do the following to the items
+    #
+    # 1. Read them from the per-seq areas
+    # 2. Write them back into the buffer
+    expected_micros_per_seq = theor_load_micros_per_seq * 2
+
+    # warmup
+    active_queries.to_garbage_padded()
+
+    torch.cuda.synchronize()
+    started_at = time.time()
+    n_iters = 10
+    for _ in range(n_iters):
+        active_queries.to_garbage_padded()
+
+    torch.cuda.synchronize()
+    elapsed_micros = (time.time() - started_at) * 1e6
+
+    micros_per_mb = elapsed_micros / n_iters
+    micros_per_seq = micros_per_mb / n_seqs
+    print(
+        f"""
+# Theoretical
+{bytes_in_queries_total/1e9=:.3f}GB
+{bytes_in_queries_per_seq/1e6=:.2f}MB
+{theor_load_micros_per_seq=:.1f}µs per seq (to just load once from memory)
+{expected_micros_per_seq=:.1f}µs per seq
+
+# Actual
+{micros_per_mb=:.1f}µs per microbatch
+{micros_per_seq=:.1f}µs per seq
+
+{micros_per_seq/expected_micros_per_seq:.1f}x the expected HBM-bandwidth bound time
+"""
+    )
+
+
+def test_calculate_scores_via_qk_dotprod_throughput(
+    n_key_ctx_per_seq=1024, n_active_query_ctx_per_seq=5
+):
+    n_seqs = 100
+    d_model_per_gpu = 12 * 1024 // 8
+    seq_kv_cache = [
+        _single_seq_kv_cache(n_ctx=n_key_ctx_per_seq, value=42, d_model=d_model_per_gpu)
+        for _ in range(n_seqs)
+    ]
+
+    active_queries = RaggedActivations.from_list(
+        [
+            torch.ones(n_active_query_ctx_per_seq, d_model_per_gpu, **bf16_cuda()) * 2
+            for _ in range(n_seqs)
+        ]
+    )
+    assert n_key_ctx_per_seq > n_active_query_ctx_per_seq * 10, (
+        "n_active_query_ctx_per_seq must be much larger than "
+        "n_key_ctx_per_seq for our simulator to be useful because "
+        "we round the HBM memory bandwidth for the active_queries and "
+        "for the scores down to zero"
+    )
+
+    bytes_in_keys_per_seq = n_key_ctx_per_seq * d_model_per_gpu * 2  # 2 from bf16
+    bytes_in_keys_total = bytes_in_keys_per_seq * n_seqs
+    hbm_bw_bytes_per_gpu = 1555e9  # 1.5TB/s
```

Comment:
not for this PR, but I would not do this (compare a number to a theoretical one), (a) it's HW specific -how does this test relate to another accelerator ?- and (b) does some suppositions on what's going on -here the data format for instance-.

For other benchmarks we calcul the user-facing throughput, it's also for instance what Phil does [here](https://github.com/openai/triton/blob/master/python/bench/bench_cross_entropy.py#L24), in that you consider the implementation as a black box, and you count the bytes going in and out (at best for instance you read the seqs and write the attention matrix, the rest is history). It's mostly what happens in this code already, but I would 
- count the BW with num elem * elem_size() (and not suppose bfloat16, would be nice to compare across types actually, it can give an idea on how the kernels are compute or bandwidth bound, at least it helped me on other tasks)

- test for a bunch of sizes, from experience there are a lot of possible holes / scheduling, and testing with one size only is like russian roulette ([see for instance](https://facebookresearch.github.io/xformers/tutorials/triton.html), there are [helpers](https://github.com/facebookresearch/xformers/blob/main/xformers/benchmarks/utils.py#L28) for this in the repo if that helps)

I can do that later on if this ends up running locally, brain dump here :)